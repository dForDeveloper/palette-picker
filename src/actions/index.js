export const setColors = (colors) => ({
  type: 'SET_COLORS',
  colors
});

export const setLockedColors = (lockedColors) => ({
  type: 'SET_LOCKED_COLORS',
  lockedColors
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const setPalettes = (id, palettes) => ({
  type: 'SET_PALETTES',
  id,
  palettes
});

export const setModal = (isDisplayed, modalType = null, text = null) => ({
  type: 'SET_MODAL',
  isDisplayed,
  modalType,
  text
});