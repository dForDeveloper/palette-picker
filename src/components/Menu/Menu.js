import React, { Component } from 'react';

export class Menu extends Component {
  handleClick = () => {
    this.props.toggleModal();
  }

  render() {
    return (
      <section className="Menu">
        <h2 className="Menu--h2">projects</h2>
        <button className="Menu--button" onClick={this.handleClick}>
          save palette
        </button>
      </section>
    );
  }
}