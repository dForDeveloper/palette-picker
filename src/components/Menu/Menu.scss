.Menu {
  width: 100%;
  background-color: #0b0b0e;
  color: #9facbd;
  display: grid;
  grid-template-rows: 1fr calc(4rem + 2px);
}

.Menu--h2 {
  text-align: center;
  font-weight: 400;
  font-size: 1.3rem;
  margin-top: 8px;
}

.Menu--button {
  height: 2rem;
  margin: 0 auto calc(2rem + 2px);
  padding: 0 16px 2px;
  border-radius: 2px;
  outline: none;
  cursor: pointer;
}