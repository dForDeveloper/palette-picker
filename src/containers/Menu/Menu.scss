.Menu {
  width: 100%;
  background-color: #0b0b0e;
  color: #9facbd;
  display: grid;
  grid-template-rows: 32px 1fr calc(4rem + 2px);
  position: relative;
}

.Menu--text {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 200px;
  display: inline-block;
  vertical-align: middle;
}

.Menu--header {
  display: grid;
  grid-template-columns: 1fr 32px;
  cursor: pointer;
}

.Menu--h2 {
  @extend .Menu--text;
  font-weight: 400;
  font-size: 1.1rem;
  margin: 8px;
  display: inline-block;
}

.Menu--ul-palettes {
  list-style-type: none;
  margin: 8px;
}

.Menu--button {
  height: 2rem;
  margin: 0 auto calc(2rem + 2px);
  padding: 0 16px 2px;
  border-radius: 2px;
  outline: none;
  cursor: pointer;
}

.Menu--icon-arrow {
  vertical-align: middle;
  width: 32px;
  padding-top: 2px;
}

.Menu--drop-down {
  position: absolute;
  background-color: #ffffff;
  top: 32px;
  width: 232px;
  left: 8px;
  list-style-type: none;
  border-radius: 8px;
  padding: 12px 4px 8px;
}