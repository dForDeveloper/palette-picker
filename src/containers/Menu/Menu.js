import React, { Component } from 'react';
import { connect } from 'react-redux';
import ProjectListItem from '../ProjectListItem/ProjectListItem';
import { setModal } from '../../actions';
import PropTypes from 'prop-types';
import PaletteListItem from '../PaletteListItem/PaletteListItem';

export class Menu extends Component {
  state = {
    activeProjectID: null,
    dropDownDisplayed: false
  };

  toggleModal = () => {
    this.props.setModal(true, 'save');
  }

  renderPaletteList = () => {
    const { palettes } = this.props;
    const { activeProjectID } = this.state;
    if (!palettes[activeProjectID]) return;
    if (activeProjectID) {
      return palettes[activeProjectID].map(palette => {
        return <PaletteListItem key={palette.id} {...palette} />
      });
    }
  };

  renderProjectList = () => {
    return this.props.projects.map(project => {
      return (
        <ProjectListItem
          key={project.name}
          {...project}
          toggleDropDown={this.toggleDropDown}
          setActiveProject={this.setActiveProject}
        />
      );
    });
  };

  toggleDropDown = () => {
    const { dropDownDisplayed } = this.state;
    this.setState({ dropDownDisplayed: !dropDownDisplayed });
  }

  setActiveProject = (id) => {
    this.setState({ activeProjectID: id });
  }

  render() {
    const { dropDownDisplayed, activeProjectID } = this.state;
    const foundProject = this.props.projects.find(project => {
      return project.id === activeProjectID;
    });
    let menuHeader = 'select a project';
    if (foundProject) {
      menuHeader = foundProject.name;
    }
    return (
      <section className="Menu">
        <header className="Menu--header" onClick={this.toggleDropDown}>
          <h2 className="Menu--h2">{menuHeader}</h2>
          <div className="Menu--icon-arrow"/>
        </header>
        <ul className="Menu--ul-palettes">{this.renderPaletteList()}</ul>
        <button className="Menu--button" onClick={this.toggleModal}>
          save palette
        </button>
        {dropDownDisplayed && 
          <ul className="Menu--drop-down">
            {this.renderProjectList()}
          </ul>}
      </section>
    );
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects,
  palettes: state.palettes
});

export const mapDispatchToProps = (dispatch) => ({
  setModal: (isDisplayed, type) => dispatch(setModal(isDisplayed, type))
});

export default connect(mapStateToProps, mapDispatchToProps)(Menu);

Menu.propTypes = {
  projects: PropTypes.array,
  palettes: PropTypes.object,
  setModal: PropTypes.func
};