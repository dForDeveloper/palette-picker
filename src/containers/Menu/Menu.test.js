import React from 'react';
import { shallow } from 'enzyme';
import { setModal } from '../../actions';
import { Menu, mapStateToProps, mapDispatchToProps } from './Menu';

describe('Menu', () => {
  let wrapper;
  const mockProps = {
    projects: [{ id: 1, name: 'my project' }],
    palettes: { '1': [{ id: 1, project_id: 1 }] },
    setModal: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<Menu {...mockProps} />);
  });

  describe('render', () => {
    it('should match the snapshot when there is no active project', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when there is an active project', () => {
      wrapper.setState({ activeProjectID: 1 });
      expect(wrapper).toMatchSnapshot();
    });
    
    it('should match the snapshot when the drop down is displayed', () => {
      wrapper.setState({ dropDownDisplayed: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('should call toggleDropDown when the header is clicked', () => {
      jest.spyOn(wrapper.instance(), 'toggleDropDown');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--header').simulate('click');
      expect(wrapper.instance().toggleDropDown).toHaveBeenCalled();
    });
    
    it('should call toggleModal when save palette is clicked', () => {
      jest.spyOn(wrapper.instance(), 'toggleModal');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--button').simulate('click');
      expect(wrapper.instance().toggleModal).toHaveBeenCalled();
    });
  });

  describe('toggleModal', () => {
    it('should call setModal(true, "save")', () => {
      wrapper.instance().toggleModal();
      expect(mockProps.setModal).toHaveBeenCalledWith(true, 'save');
    });
  });

  describe('renderPaletteList', () => {
    it('should return undefined if there are no palettes', () => {
      const result = wrapper.instance().renderPaletteList();
      expect(result).toEqual(undefined);
    });

    it('should return an array equal in length to the palettes array', () => {
      wrapper.setState({ activeProjectID: 1 });
      const result = wrapper.instance().renderPaletteList();
      expect(result).toHaveLength(mockProps.palettes[1].length);
    });
  });

  describe('renderProjectList', () => {
    it('should return an array of length equal to the projects array', () => {
      const result = wrapper.instance().renderProjectList();
      expect(result).toHaveLength(mockProps.projects.length);
    });
  });

  describe('toggleDropDown', () => {
    it('should toggle dropDownDisplayed in state', () => {
      expect(wrapper.state('dropDownDisplayed')).toEqual(false);
      wrapper.instance().toggleDropDown();
      expect(wrapper.state('dropDownDisplayed')).toEqual(true);
    });
  });

  describe('setActiveProject', () => {
    it('should set activeProjectID in state', () => {
      expect(wrapper.state('activeProjectID')).toEqual(null);
      wrapper.instance().setActiveProject(1);
      expect(wrapper.state('activeProjectID')).toEqual(1);
    })
  });
});

describe('mapStateToProps', () => {
  it('should return a props object with the correct properties', () => {
    const mockState = {
      projects: [{ id: 1, name: 'my project' }],
      palettes: { '1': [{ id: 1, project_id: 1 }] },
      extraProperty: true
    };
    const expected = {
      projects: [{ id: 1, name: 'my project' }],
      palettes: { '1': [{ id: 1, project_id: 1 }] }
    };
    const result = mapStateToProps(mockState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  it('should call setModal', () => {
    const mockDispatch = jest.fn();
    const mappedProps = mapDispatchToProps(mockDispatch);
    const expected = setModal(true, 'save');
    mappedProps.setModal(true, 'save');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});