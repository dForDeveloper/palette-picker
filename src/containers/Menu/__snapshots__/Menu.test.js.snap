// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu render should match the snapshot when the drop down is displayed 1`] = `
<section
  className="Menu"
>
  <header
    className="Menu--header"
    onClick={[Function]}
  >
    <h2
      className="Menu--h2"
    >
      select a project
    </h2>
    <div
      className="Menu--icon-arrow"
    />
  </header>
  <ul
    className="Menu--ul-palettes"
  />
  <button
    className="Menu--button"
    onClick={[Function]}
  >
    save palette
  </button>
  <ul
    className="Menu--drop-down"
  >
    <Connect(ProjectListItem)
      id={1}
      key="my project"
      name="my project"
      setActiveProject={[Function]}
      toggleDropDown={[Function]}
    />
  </ul>
</section>
`;

exports[`Menu render should match the snapshot when there is an active project 1`] = `
<section
  className="Menu"
>
  <header
    className="Menu--header"
    onClick={[Function]}
  >
    <h2
      className="Menu--h2"
    >
      my project
    </h2>
    <div
      className="Menu--icon-arrow"
    />
  </header>
  <ul
    className="Menu--ul-palettes"
  >
    <Connect(PaletteListItem)
      id={1}
      key="1"
      project_id={1}
    />
  </ul>
  <button
    className="Menu--button"
    onClick={[Function]}
  >
    save palette
  </button>
</section>
`;

exports[`Menu render should match the snapshot when there is no active project 1`] = `
<section
  className="Menu"
>
  <header
    className="Menu--header"
    onClick={[Function]}
  >
    <h2
      className="Menu--h2"
    >
      select a project
    </h2>
    <div
      className="Menu--icon-arrow"
    />
  </header>
  <ul
    className="Menu--ul-palettes"
  />
  <button
    className="Menu--button"
    onClick={[Function]}
  >
    save palette
  </button>
</section>
`;
